storage/streams/src/main/java/io/apicurio/registry/streams/distore/DistributedReadOnlyKeyValueStore.java

```diff
@@ -0,0 +1,109 @@
+ * the distributed streams application. It dispatches requests to stores on local and remote
+ * KafkaStreams processing nodes that contain parts of the data which is looked up.
+ *
+ * @param <K> the type of keys
+ * @param <V> the type of values
+ */
+public class DistributedReadOnlyKeyValueStore<K, V>
+    extends DistributedReadOnlyStateStore<K, V, ExtReadOnlyKeyValueStore<K, V>>
+    implements ExtReadOnlyKeyValueStore<K, V> {
+
+    /**
+     * @param streams                The {@link KafkaStreams} application
+     * @param localApplicationServer The {@link HostInfo} derived from the
+     *                               {@link StreamsConfig#APPLICATION_SERVER_CONFIG application.server}
+     *                               configuration property of local kafka streams node for the streams application.
+     *                               This is used to identify requests for local store, bypassing gRPC calls
+     * @param storeName              The name of the {@link ReadOnlyKeyValueStore} registered in the streams application
+     * @param keySerde               The {@link Serde} for keys of the store
+     * @param valSerde               The {@link Serde} for values of the store
+     * @param grpcChannelProvider    A function that establishes gRPC {@link Channel} to a remote store service
+     *                               for the given {@link HostInfo} parameter
+     * @param parallel               {@code true} if lookups that need to query many stores in the cluster are
+     *                               to be performed in parallel
+     */
+    public DistributedReadOnlyKeyValueStore(
+        KafkaStreams streams,
+        HostInfo localApplicationServer,
+        String storeName,
+        Serde<K> keySerde, Serde<V> valSerde,
+        Function<? super HostInfo, ? extends Channel> grpcChannelProvider,
+        boolean parallel
+    ) {
+        super(
+            streams,
+            localApplicationServer,
+            storeName,
+            keySerde,
+            valSerde,
+            grpcChannelProvider,
+            parallel
+        );
+    }
+
+    @Override
+    protected ExtReadOnlyKeyValueStore<K, V> localService(String storeName, KafkaStreams streams) {
+        ReadOnlyKeyValueStore<K, V> delegate = streams.store(storeName, QueryableStoreTypes.keyValueStore());
+        return new ExtReadOnlyKeyValueStoreImpl<>(delegate);
+    }
+
+    @Override
+    protected ExtReadOnlyKeyValueStore<K, V> remoteServiceGrpcClient(String storeName, Channel channel, Serde<K> keySerde, Serde<V> valSerde) {
+        return new ReadOnlyKeyValueStoreGrpcClient<>(storeName, channel, keySerde, valSerde);
+    }
+
+    @Override
+    public CloseableIterator<K> allKeys() {
+        return new StreamToIteratorAdapter<K>(
+            allServicesStream().flatMap(store -> StreamToIteratorAdapter.toStream(store.allKeys()))
+        );
+    }
```

Comment:
So is it possible for this `allKeys` method to return duplicates? I.e. if there could be multiple services which each (due to repartitioning for example) had a given key. Maybe the caller(s) of this method ultimately don't care?

Reply:
The keys are properly split among nodes -- if all is OK.
And I guess when repartitioning is on-going, I think you don't get StreamsMetadata.
Or I could be wrong?

>> Maybe the caller(s) of this method ultimately don't care?

Good question. It probably depends on the use-case ... ?

Reply:
> I think you don't get StreamsMetadata.
Or I could be wrong?

IDK either. The Javadoc says

> Note: this is a point in time view and it may change due to partition reassignment.Note: this is a point in time view and it may change due to partition reassignment.

which just says to me that you can expect this to change between reassignments, leaving undefined how it behaves _during_ a reassignment. Having taken a very quick look at the code it looks to me like you can call it during rebalance and it probably is consistent. So let's not worry too much about this.

Reply:
Yeah ... plus in our case, we currently use this to return all artifac ids ... so not really a problem, imo.

When / if we make this a generic library (the distore / diservice code) one day, we can revisit this.